src/lib/common/LineMeasureController.svelte

```diff
@@ -0,0 +1,167 @@
+    const maxLngLatDistance = new LngLat(0, -90).distanceTo(new LngLat(0, 90));
+    let currentShortestDistance = maxLngLatDistance;
+    let waypointToRemove: LngLat | undefined = undefined;
+
+    waypoints.forEach((waypoint) => {
+      const waypointLngLat = waypoint.lngLat;
+      if (currentShortestDistance >= clickLocation.distanceTo(waypointLngLat)) {
+        waypointToRemove = waypointLngLat;
+        currentShortestDistance = clickLocation.distanceTo(waypointLngLat);
+      }
+    });
+
+    if (waypointToRemove) {
+      waypoints = waypoints.filter((waypoint) => {
+        const waypointLngLat = waypoint.lngLat;
+        return (
+          waypointLngLat.lng != waypointToRemove?.lng ||
+          waypointLngLat.lat != waypointToRemove?.lat
+        );
+      });
+    }
+  }
+
+  function enableMeasurement() {
+    isInactive = false;
+    $map.on("click", handleMapClickEvent);
+  }
+
+  function disableMeasurement() {
+    isInactive = true;
+    $map.off("click", handleMapClickEvent);
+    waypoints = [];
+    removeLineFromMap();
+  }
+
+  function keyDown(keyDownEvent: KeyboardEvent) {
+    if (keyDownEvent.key === "Shift") isShiftDown = true;
+  }
+
+  function keyUp(keyUpEvent: KeyboardEvent) {
+    if (keyUpEvent.key === "Shift") isShiftDown = false;
+  }
+
+  function waypointsUpdated() {
+    if (waypoints.length < 2) {
+      removeLineFromMap();
+      return;
+    }
+
+    const waypointsAsPosition = waypoints.map((waypoint) => {
+      return [waypoint.lngLat.lng, waypoint.lngLat.lat];
+    });
+    lineToMeasure = lineString(waypointsAsPosition, {});
+    if (lineToMeasure.properties)
+      lineToMeasure.properties.length = length(lineToMeasure) * 1000;
+    updateLinestring();
+  }
+
+  function removeLineFromMap() {
+    if ($map.getSource(layerName)) {
+      $map.removeLayer(layerName);
+      $map.removeSource(layerName);
+    }
+  }
+
+  function updateLinestring() {
+    if (!$map.getSource(layerName)) {
+      $map.addSource(layerName, {
+        type: "geojson",
+        data: lineToMeasure,
+      });
+
+      $map.addLayer({
+        id: layerName,
+        type: "line",
+        source: layerName,
+        paint: {
+          "line-color": colors.measuringLine,
+          "line-width": 5,
+        },
+      });
+    } else {
+      // @ts-ignore setData not happy for some reason
+      $map.getSource(layerName)?.setData(lineToMeasure);
+    }
+  }
+</script>
+
+{#if isInactive}
+  <SecondaryButton on:click={enableMeasurement}>
+    Enable Line Measurement Tool
+  </SecondaryButton>
+{:else}
+  <SecondaryButton on:click={disableMeasurement}>
+    Disable Line Measurement Tool
+  </SecondaryButton>
+{/if}
+{#if !isInactive}
+  {#if lineToMeasure}
+    <p>Length: {lineToMeasure.properties?.length}m</p>
+  {/if}
+  <CollapsibleCard label="Help">
+    <ul>
+      <li>
+        <b>Click</b>
+        on the map to place a new waypoint
+      </li>
+      <li>
+        <b>Click & Drag</b>
```

Comment:
"and drag"

btw "&" isn't supposed to be used directly in HTML, since it's used for special codes. &amp;